span_context.go

```diff
@@ -82,6 +83,10 @@ type samplingState struct {
 	// When state is not final, sampling will be retried on other span write operations,
 	// like SetOperationName / SetTag, and the spans will remain writable.
 	final atomic.Bool
+
+	// extendedState allows samplers to keep intermediate state.
+	// The keys and values in this map are completely opaque: interface{} -> interface{}.
+	extendedState sync.Map
```

Comment:
I'm concerned that we are prematurely trading away type safety for performance - do we have quantifications of the gains? 

The doc says the following - 
> It is optimized for use in concurrent loops with keys that are stable over time, and either few steady-state stores, or stores localized to one goroutine per key.
> For use cases that do not share these attributes, it will likely have comparable or worse performance and worse type safety than an ordinary map paired with a read-write mutex.

Reply:
I would like to do benchmarks, first I wanted to get the code in a working state with unit tests, so that if we start doing refactoring & perf optimizations there's a foundation.  

We definitely need race-safety here, so the alternative would be using map with a lock, which actually involves more code than sync.Map (which has its drawbacks and benchmarks may demonstrate that)